enumerable_methods.rb

```diff
@@ -0,0 +1,182 @@
+    instance = instance.clone
+    return instance.to_enum unless block_given?
+
+    i = 0
+    loop do
+      yield(instance[i])
+      i += 1
+      break if i == instance.length
+    end
+  end
+
+  def my_each_with_index
+    instance = self
+    instance = instance.clone
+    return instance.to_enum unless block_given?
+
+    i = 0
+    loop do
+      yield(instance[i], i)
+      i += 1
+      break if i == instance.length
+    end
+  end
+
+  def my_select
+    instance = self
+    instance = instance.clone
+    result = []
+    return instance.to_enum unless block_given?
+
+    instance.my_each do |e|
+      result.push(e) if yield(e)
+    end
+    result
+  end
+
+  def my_all?(type = nil)
+    instance = self
+    instance = instance.clone
+    puts 'warning: given block not used' if block_given? && !type.nil?
+    instance.my_each do |e|
+      is_true = truthy(e, type)
+      is_true ||= truthy_no_typeblock(e, type.nil?, !block_given?)
+      return false unless is_true
+
+      if block_given?
+        return false unless yield(e)
+      end
+    end
+    true
+  end
+
+  def my_any?(type = nil)
+    instance = self
+    instance = instance.clone
+    puts 'warning: given block not used' if block_given? && !type.nil?
+    !instance.my_none?(type)
+  end
+
+  def my_none?(type = nil)
+    instance = self
+    instance = instance.clone
+    puts 'warning: given block not used' if block_given? && !type.nil?
+    instance.my_each do |e|
+      is_true = truthy(e, type)
+      is_true ||= truthy_no_typeblock(e, type.nil?, !block_given?)
+      return false if is_true
+      return false if block_given? && yield(e)
+    end
+    true
+  end
+
+  def my_count(arg = nil, &proc)
+    instance = self
+    instance = instance.clone
+    if !arg.nil?
+      puts 'warning: given block not used' unless proc.nil?
+      count = 0
+      instance.my_each do |e|
+        count += 1 if e == arg
+      end
+      count
+    elsif !proc.nil?
+      count = 0
+      instance.my_each do |e|
+        count += 1 if proc.call(e)
+      end
+      count
+    else
+      instance.length
+    end
+  end
+
+  def my_map
+    instance = self
+    instance = instance.clone
+    result = []
+    return instance.to_enum unless block_given?
+
+    instance.my_each do |e|
+      result.push(yield(e))
+    end
+    result
+  end
+
+  def my_inject(initial = nil, sym = nil)
+    operations = {
+      :+ => proc { |a, b| a + b },
+      :- => proc { |a, b| a - b },
+      :* => proc { |a, b| a * b },
+      :/ => proc { |a, b| a / b },
+      :** => proc { |a, b| a**b }
+    }
+    instance = self
+    instance = instance.clone
+    if sym.nil? && initial.respond_to?(:to_sym)
+      sym = initial.to_sym
+      initial = nil
+    end
+    total = initial
+    total = instance.shift if initial.nil?
+    instance.my_each do |e|
+      total = operations[sym].call(total, e) unless sym.nil?
+      total = yield(total, e) if block_given?
+    end
+    total
+  end
```

Comment:
`p (5..10).my_inject { |sum, n| sum + n }` returns `error` instead of `45`
`p (5..10).my_inject(1) { |product, n| product * n }` returns `error` instead of `151200`